op-node/sources/fallback_client.go

```diff
@@ -0,0 +1,243 @@
+	"sync/atomic"
+	"time"
+)
+
+// FallbackClient is an RPC client, it can automatically switch to the next l1 endpoint
+// when there is a problem with the current l1 endpoint
+// and automatically switch back after the first l1 endpoint recovers.
+type FallbackClient struct {
+	// firstRpc is created by the first of the l1 urls, it should be used first in a healthy state
+	firstRpc          client.RPC
+	urlList           []string
+	rpcInitFunc       func(url string) (client.RPC, error)
+	lastMinuteFail    atomic.Int64
+	currentRpc        client.RPC
+	currentIndex      int
+	mx                sync.Mutex
+	log               log.Logger
+	isInFallbackState bool
+	subscribeFunc     func() (event.Subscription, error)
+	l1HeadsSub        *ethereum.Subscription
+	l1ChainId         *big.Int
+	l1Block           eth.BlockID
+	ctx               context.Context
+}
+
+// NewFallbackClient returns a new FallbackClient. l1ChainId and l1Block are used to check
+// whether the newly switched rpc is legal.
+func NewFallbackClient(ctx context.Context, rpc client.RPC, urlList []string, log log.Logger, l1ChainId *big.Int, l1Block eth.BlockID, rpcInitFunc func(url string) (client.RPC, error)) client.RPC {
+	fallbackClient := &FallbackClient{
+		ctx:          ctx,
+		firstRpc:     rpc,
+		urlList:      urlList,
+		log:          log,
+		rpcInitFunc:  rpcInitFunc,
+		currentRpc:   rpc,
+		currentIndex: 0,
+		l1ChainId:    l1ChainId,
+		l1Block:      l1Block,
+	}
+	go func() {
+		ticker := time.NewTicker(1 * time.Minute)
+		for {
+			select {
+			case <-ticker.C:
+				log.Debug("FallbackClient clear lastMinuteFail 0")
+				fallbackClient.lastMinuteFail.Store(0)
+			}
+		}
+	}()
+	return fallbackClient
+}
+
+func (l *FallbackClient) SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error) {
+	return l.EthSubscribe(ctx, ch, "newHeads")
+}
+
+func (l *FallbackClient) Close() {
+	l.currentRpc.Close()
+	if l.currentRpc != l.firstRpc {
+		l.firstRpc.Close()
+	}
+}
+
+func (l *FallbackClient) CallContext(ctx context.Context, result any, method string, args ...any) error {
+	err := l.currentRpc.CallContext(ctx, result, method, args...)
+	if err != nil {
+		l.handleErr(err)
+	}
+	return err
+}
+
+func (l *FallbackClient) handleErr(err error) {
+	if err == rpc.ErrNoResult {
+		return
+	}
+	errCount := l.lastMinuteFail.Add(1)
+	if errCount > 10 {
+		l.switchCurrentRpc()
+	}
+}
+
+func (l *FallbackClient) BatchCallContext(ctx context.Context, b []rpc.BatchElem) error {
+	err := l.currentRpc.BatchCallContext(ctx, b)
+	if err != nil {
+		l.handleErr(err)
+	}
+	return err
+}
+
+func (l *FallbackClient) EthSubscribe(ctx context.Context, channel any, args ...any) (ethereum.Subscription, error) {
+	subscribe, err := l.currentRpc.EthSubscribe(ctx, channel, args...)
+	if err != nil {
+		l.handleErr(err)
+	}
+	return subscribe, err
+}
+
+func (l *FallbackClient) switchCurrentRpc() {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+	if l.lastMinuteFail.Load() <= 10 {
```

Comment:
The `10` here is hardcoded, while in the implementation of the similar function, there is a `fallbackThreshold` parameter.

Reply:
There is no need for parameters in op-node, because it is a fixed value from beginning to end. It may be better to add a flag to specify for the user. But we need to minimize the increase of flags. The magic value is really bad, I've made it a const